src/rulesets/LinearAlgebra/norm.jl

```diff
@@ -17,20 +17,25 @@ function frule((_, Δx), ::typeof(norm), x::Number, p::Real)
     return y, ∂y
 end
 
-function rrule(
-    ::typeof(norm),
-    x::Union{StridedArray, LinearAlgebra.AbstractTriangular, Diagonal},
-    p::Real,
-)
+function rrule(::typeof(norm), x::AbstractArray, p::Real)
     y = LinearAlgebra.norm(x, p)
-    function norm_pullback(Δy)
+    function norm_pullback_p(Δy)
         ∂x = Thunk() do
             return if isempty(x) || p == 0
-                zero.(x) .* (zero(y) * zero(real(Δy)))
+                InplaceableThunk(
```

Comment:
Putting an `InplacableThunk` inside a `Thunk` is weird.
I would expect it to break things, since they probably assume that they can `unthunk` once to remove all thunks.

Reply:
OK, not sure how I missed that. 

Is it OK to return a different `InplaceableThunk` for different values of `p`? Otherwise the branching on `p` could go inside, but I think it will be a bit messy as you'll have to repeat it for out-of-place and in-place variants.

Reply:
Or the branching could be further outside, like
```
norm_pullback_p = if isempty(x) || p == 0
    Δy -> InplaceableThunk(...)
elseif p == 2
    Δy -> InplaceableThunk(...)
...
```

Reply:
Maybe leading to
```
125
  Got exception outside of a @test
126
  return type Tuple{Zero, Vector{Float64}} does not match inferred return type Tuple{Zero, Union{InplaceableThunk{Thunk{ChainRules.var"#1780#1783"{Float64, Vector{Float64}, Float64}}, ChainRules.var"#1781#1784"{Float64, Vector{Float64}, Float64}}, Vector{Float64}}}
127
  Stacktrace:
143
      @ ~/work/ChainRules.jl/ChainRules.jl/test/rulesets/LinearAlgebra/norm.jl:75 [inlined]
```

Reply:
OK, done in https://github.com/JuliaDiff/ChainRules.jl/pull/337/commits/15e55eef3df10c8c8b256382423e2479f4ee0a27

Reply:
Now I think it gives other errors
```
return type Tuple{Zero,InplaceableThunk{Thunk{getfield(ChainRules, Symbol("##1770#1783")){Float64,Array{Float64,3},Float64}},getfield(ChainRules, Symbol("##1775#1788")){Float64,Array{Float64,3},Float64}},Thunk{getfield(ChainRules, Symbol("##1778#1791")){Float64,Array{Float64,3},Float64,Float64}}} does not match inferred return type Tuple{Zero,InplaceableThunk{_1,_2} where _2 where _1,Thunk{getfield(ChainRules, Symbol("##1778#1791")){Float64,Array{Float64,3},Float64,Float64}}}
```

Reply:
I've removed these. But I get errors like this and have no idea what they mean, nor where they come from:

<img width="788" alt="Screenshot 2021-04-29 at 14 43 06" src="https://user-images.githubusercontent.com/32575566/116612356-7226ad00-a905-11eb-99cd-33b809b33f33.png">

Reply:
Oh at last I found out. Two arguments two thunks is fine. One argument one thunk is, for some reason, forbidden by a test. Why? (Disabled for now.)

Reply:
We assume that if someone is calling the pullback that they want to actually do AD and find at least one derivative.
The point of thunking is as a poorman's version of Haskell (and TensorFlow's) lazyness where nothing not asked for gets run.
But since this is the only output* we can assume that it will be asked for.
So the indirection and complication of thunking things isn't benificial.
https://juliadiff.org/ChainRulesCore.jl/dev/writing_good_rules.html#Use-Thunks-appropriately

* ignoring functors.

Reply:
OK, but this sounds a bit like making `2^1` an error -- there is nothing gained by writing this, but nothing lost, and if you make it an error then it's one more confusing edge case to keep track of. For example, if you want to comment out one argument of a function temporarily, your tests will now fail.

Reply:
If you must, you could print a warning stating that this is poor style. But if it's not actually wrong, then it's irritating that an hour-long run of tests gets derailed by this... it's plenty hard enough already to make things pass.

Reply:
Fixed by https://github.com/JuliaDiff/ChainRulesTestUtils.jl/pull/144